kernel/sched.c

```diff
@@ -1338,44 +1338,49 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
 }
 
 #ifdef CONFIG_SCHED_MULTIQ
-# if (K_LOWEST_THREAD_PRIO - K_HIGHEST_THREAD_PRIO) > 31
-# error Too many priorities for multiqueue scheduler (max 32)
-# endif
 
 static ALWAYS_INLINE void z_priq_mq_add(struct _priq_mq *pq,
 					struct k_thread *thread)
 {
-	int priority_bit = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int offset_priority = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int priority_idx = BIT_ARRAY_IDX(offset_priority);
+	int priority_bit = BIT_ARRAY_OFFSET(offset_priority);
 
-	sys_dlist_append(&pq->queues[priority_bit], &thread->base.qnode_dlist);
-	pq->bitmask |= BIT(priority_bit);
+	sys_dlist_append(&pq->queues[offset_priority], &thread->base.qnode_dlist);
+	pq->bitmask[priority_idx] |= BIT(priority_bit);
 }
 
 static ALWAYS_INLINE void z_priq_mq_remove(struct _priq_mq *pq,
 					   struct k_thread *thread)
 {
-	int priority_bit = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int offset_priority = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int priority_idx = BIT_ARRAY_IDX(offset_priority);
+	int priority_bit = BIT_ARRAY_OFFSET(offset_priority);
 
 	sys_dlist_remove(&thread->base.qnode_dlist);
-	if (sys_dlist_is_empty(&pq->queues[priority_bit])) {
-		pq->bitmask &= ~BIT(priority_bit);
+	if (sys_dlist_is_empty(&pq->queues[offset_priority])) {
+		pq->bitmask[priority_idx] &= ~BIT(priority_bit);
 	}
 }
 #endif
 
 struct k_thread *z_priq_mq_best(struct _priq_mq *pq)
 {
-	if (!pq->bitmask) {
-		return NULL;
-	}
-
 	struct k_thread *thread = NULL;
-	sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask)];
-	sys_dnode_t *n = sys_dlist_peek_head(l);
 
-	if (n != NULL) {
-		thread = CONTAINER_OF(n, struct k_thread, base.qnode_dlist);
+	for (int i = 0; i < PRI_BITMAP_ITERATOR; ++i) {
+		if (!pq->bitmask[i]) {
+			continue;
+		}
+
+		sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask[i])];
+		sys_dnode_t *n = sys_dlist_peek_head(l);
+
+		if (n != NULL) {
+			thread = CONTAINER_OF(n, struct k_thread, base.qnode_dlist);
```

Comment:
This logic looks wrong?  If you find a thread in the first word, you'll continue searching in the other words (which hold lower priorities!) and return the wrong thread.  You want to return right here, no?

Reply:
You're right. It's wrong. I will fixed it later.